File: api/src/main/java/org/pmiops/workbench/elasticsearch/AggregationUtils.java
package org.pmiops.workbench.elasticsearch;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import org.elasticsearch.action.search.SearchResponse;
import org.elasticsearch.search.aggregations.AggregationBuilder;
import org.elasticsearch.search.aggregations.AggregationBuilders;
import org.elasticsearch.search.aggregations.BucketOrder;
import org.elasticsearch.search.aggregations.bucket.range.ParsedDateRange;
import org.elasticsearch.search.aggregations.bucket.range.Range;
import org.elasticsearch.search.aggregations.bucket.terms.Terms;
import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
import org.pmiops.workbench.model.AgeType;
import org.pmiops.workbench.model.DemoChartInfo;
import org.pmiops.workbench.model.GenderOrSexType;

public class AggregationUtils {

  public static final String RANGE_18_44 = "18-44";
  public static final String RANGE_45_64 = "45-64";
  public static final String RANGE_GT_65 = "65";
  public static final String DATE = "d_";
  public static final String GENDER_OR_SEX = "g_or_s_";
  public static final String RACE = "r_";

  /**
   * Build aggregations for demographic charting based on date range, gender and race. The bucket
   * aggregations compute and return the number of documents that "fell into" each bucket.
   */
  public static AggregationBuilder buildDemoChartAggregation(
      GenderOrSexType genderOrSexType, AgeType ageType, String ageRange) {
    String[] ages = ageRange.split("-");
    // use Integer values of age ranges when calculating age at consent or age at cdr
    Integer lo = Integer.valueOf(ages[0]);
    Integer hi = (ages.length > 1) ? Integer.valueOf(ages[1]) : null;

    // use high endDate of the age range to calculate the startDate of the date range
    // need to add 1 year to adjust the startDate date to beginning of age range
    LocalDate startDate = (hi != null) ? ElasticUtils.todayMinusYears(hi + 1) : null;
    // use the low endDate of the age range to calculate the endDate of the date range
    // when calculating age from date of birth
    LocalDate endDate = ElasticUtils.todayMinusYears(lo);

    boolean isGender = GenderOrSexType.GENDER.equals(genderOrSexType);
    boolean isAgeAtConsent = AgeType.AGE_AT_CONSENT.equals(ageType);
    // Added order to gender/sex and race buckets. Therefore the UI code can expect consistent
    // results between BQ(sql ordered by gender/sex, race, age) and elastic
    TermsAggregationBuilder termsAggregationBuilder =
        AggregationBuilders.terms(GENDER_OR_SEX + ageRange)
            .field(isGender ? "gender_concept_name" : "sex_at_birth_concept_name")
            .order(BucketOrder.key(true))
            .subAggregation(
                AggregationBuilders.terms(RACE + ageRange)
                    .field("race_concept_name")
                    .order(BucketOrder.key(true))
                    // This eliminates the race buckets with 0 counts. Without this param
                    // elastic returns all race buckets regardless of count.
                    .minDocCount(1));
    return AgeType.AGE.equals(ageType)
        ? AggregationBuilders.dateRange(DATE + ageRange)
            .field("birth_datetime")
            .format("yyyy-MM-dd")
            .addRange((startDate == null) ? null : startDate.toString(), endDate.toString())
            .subAggregation(termsAggregationBuilder)
        : AggregationBuilders.range(DATE + ageRange)
            .field(isAgeAtConsent ? "age_at_consent" : "age_at_cdr")
            .addRange(new Double(lo), new Double(hi))
            .subAggregation(termsAggregationBuilder);
  }

  public static List<DemoChartInfo> unwrapDemoChartBuckets(
      SearchResponse searchResponse, String... ageRanges) {
    List<DemoChartInfo> demoInformation = new ArrayList<>();
    for (String ageRange : ageRanges) {
      ParsedDateRange parsedDate = searchResponse.getAggregations().get(DATE + ageRange);
      for (Range.Bucket dateBucket : parsedDate.getBuckets()) {
        Terms gender = dateBucket.getAggregations().get(GENDER_OR_SEX + ageRange);
        for (Terms.Bucket genderBucket : gender.getBuckets()) {
          Terms race = genderBucket.getAggregations().get(RACE + ageRange);
          for (Terms.Bucket raceBucket : race.getBuckets()) {
            demoInformation.add(
                new DemoChartInfo()
                    .name(genderBucket.getKeyAsString().substring(0, 1))
                    .race(raceBucket.getKeyAsString())
                    .ageRange(RANGE_GT_65.equals(ageRange) ? "> " + ageRange : ageRange)
                    .count(raceBucket.getDocCount()));
          }
        }
      }
    }
    return demoInformation;
  }
}

File: api/src/main/java/org/pmiops/workbench/elasticsearch/AggregationUtils.kt
package org.pmiops.workbench.elasticsearch

import java.time.LocalDate
import java.util.ArrayList
import org.elasticsearch.action.search.SearchResponse
import org.elasticsearch.search.aggregations.AggregationBuilder
import org.elasticsearch.search.aggregations.AggregationBuilders
import org.elasticsearch.search.aggregations.BucketOrder
import org.elasticsearch.search.aggregations.bucket.range.ParsedDateRange
import org.elasticsearch.search.aggregations.bucket.range.Range
import org.elasticsearch.search.aggregations.bucket.terms.Terms
import org.pmiops.workbench.model.DemoChartInfo

object AggregationUtils {

    val RANGE_19_44 = "19-44"
    val RANGE_45_64 = "45-64"
    val RANGE_GT_65 = "65"
    val DATE = "d_"
    val GENDER = "g_"
    val RACE = "r_"

    /**
     * Build aggregations for demographic charting based on date range, gender and race. The bucket
     * aggregations compute and return the number of documents that "fell into" each bucket.
     */
    fun buildDemoChartAggregation(ageRange: String): AggregationBuilder {
        val ages = ageRange.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var start: LocalDate? = null
        // use the low end of the age range to calculate the end of the date range
        val end = ElasticUtils.todayMinusYears(Integer.parseInt(ages[0]))
        if (ages.size > 1) {
            // use high end of the age range to calculate the start of the date range
            // need to add 1 year to adjust the start date to beginning of age range
            start = ElasticUtils.todayMinusYears(Integer.parseInt(ages[1]) + 1)
        }

        // Added order to gender and race buckets. Therefore the UI code can expect consistent results
        // between BQ(sql ordered by gender, race, age) and elastic
        return AggregationBuilders.dateRange(DATE + ageRange)
                .field("birth_datetime")
                .format("yyyy-MM-dd")
                .addRange(start?.toString(), end.toString())
                .subAggregation(
                        AggregationBuilders.terms(GENDER + ageRange)
                                .field("gender_concept_name")
                                .order(BucketOrder.key(true))
                                .subAggregation(
                                        AggregationBuilders.terms(RACE + ageRange)
                                                .field("race_concept_name")
                                                .order(BucketOrder.key(true))
                                                // This eliminates the race buckets with 0 counts. Without this param
                                                // elastic
                                                // returns all race buckets regardless of count.
                                                .minDocCount(1)))
    }

    fun unwrapDemoChartBuckets(
            searchResponse: SearchResponse, vararg ageRanges: String): List<DemoChartInfo> {
        val demoInformation = ArrayList<DemoChartInfo>()
        for (ageRange in ageRanges) {
            val parsedDate = searchResponse.aggregations.get<ParsedDateRange>(DATE + ageRange)
            for (dateBucket in parsedDate.buckets) {
                val gender = dateBucket.aggregations.get<Terms>(GENDER + ageRange)
                for (genderBucket in gender.buckets) {
                    val race = genderBucket.aggregations.get<Terms>(RACE + ageRange)
                    for (raceBucket in race.buckets) {
                        demoInformation.add(
                                DemoChartInfo()
                                        .gender(genderBucket.keyAsString.substring(0, 1))
                                        .race(raceBucket.keyAsString)
                                        .ageRange(if (RANGE_GT_65 == ageRange) "> $ageRange" else ageRange)
                                        .count(raceBucket.docCount))
                    }
                }
            }
        }
        return demoInformation
    }
}
